components/raftstore/src/store/fsm/peer.rs

```diff
@@ -2408,13 +2408,16 @@ where
     fn propose_locks_before_transfer_leader(&mut self) -> bool {
         // 1. Disable in-memory pessimistic locks.
         let mut pessimistic_locks = self.fsm.peer.txn_ext.pessimistic_locks.write();
-        // If `is_valid` is false, the locks should have been proposed. But we still need to
-        // return true to propose another TransferLeader command. Otherwise, some write requests
-        // that have marked some locks as deleted will fail because raft rejects more proposals.
-        if !pessimistic_locks.is_valid {
+        // If it is not writable, it's probably because it's a retried TransferLeader and the locks
+        // have been proposed. But we still need to return true to propose another TransferLeader
+        // command. Otherwise, some write requests that have marked some locks as deleted will fail
+        // because raft rejects more proposals.
+        // It is OK to return true here if it's in other states like MergingRegion or NotLeader.
+        // In those cases, the locks will fail to propose and nothing will happen.
+        if !pessimistic_locks.is_writable() {
             return true;
         }
-        pessimistic_locks.is_valid = false;
+        pessimistic_locks.status = LocksStatus::TransferingLeader;
```

Comment:
Should we also revert the `TransferingLeader` status if the transfer is aborted? Abortion can happen in a few places (see `abort_leader_transfer` in `raft-rs`)  but we don't have a unified way to get notified...

Reply:
Yes, absolutely. I don't know how to get notified, so I decide o use a tick to do so in https://github.com/tikv/tikv/pull/11883...